ansible/library/bgp_route.py

```diff
@@ -0,0 +1,286 @@
+   *> 192.168.10.129/32
+                       10.0.0.34                              0 65200 62011 65509 ?
+   *> 192.168.10.144/32
+                       10.0.0.34                              0 65200 62011 65510 ?
+
+'''
+
+
+### TODO: ipv6 route entries parsing is not fully working, need continue working on ipv6 ###
+
+
+class BgpRoutes(object):
+    '''
+        parsing bgp routing information
+    '''
+    def __init__(self):
+        self.facts = defaultdict(dict)
+        return
+
+    def get_facts(self):
+        return self.facts
+
+    def parse_bgp_rt_adv(self, cmd, cmd_result):
+        '''
+        parse BGP routing facts of neighbor advertised routes
+        '''
+        regex_neighbor = re.compile('nei.*\s([0-9a-fA-F.:]+)\s.*')
+        #regix_iprange = re.compile('[0-9a-fA-F.:]+\/\d+')
+        if regex_neighbor.match(cmd):
+            neighbor = regex_neighbor.match(cmd).group(1)
+        else:
+            raise Exception('cannot find neighbor in "show ip  bgp ' + cmd + '"')
+        self.facts['bgp_rt_neiadv']['neighbor'] = neighbor
+        ### so far parsing by fixed length table
+        ### will improve this if find better way to parse it
+        header_position = [0, 3, 20, 40, 47, 54, 61]
+        header = 'Network          Next Hop            Metric LocPrf Weight Path'
+        result_lines = cmd_result.split('\n')
+        table_start = False
+        while len(result_lines) != 0:
+            line = result_lines.pop(0)
+            if not table_start:
+                if header in line:
+                    table_start = True
+                    continue
+            else:
+                if not re.match('^\*', line):
+                    continue
+                entry = {}
+                if len(line) > 30:   ## route entry in one line
+                    prefix = line[header_position[1]:header_position[2]].strip()
+                else:    ### route entry in two lines
+                    prefix = line[header_position[1]:30].strip()
+                    line = result_lines.pop(0)
+                entry['next_hop'] = line[header_position[2]:header_position[3]].strip()
+                paths = line[header_position[6]:].strip()
+                entry['paths'] = re.findall('[0-9]+', paths)
+                self.facts['bgp_rt_neiadv'][prefix] = entry
+
+
+    def parse_bgp_rt_prefix(self, prefix, cmd_result):
+        '''
+        parse BGP facts for specific prefix
+        '''
+        regex_prefix_header = re.compile('BGP routing table entry for ')
+        regex_prefix_avail_paths = re.compile('Paths:\s+\((\d+)\s+available')
+        prefix_temp_line1 = 'Advertised to non peer-group peers:'
+        regex_prefix_prop2 = re.compile('^[0-9a-fA-F.:\s]+$')
+        regex_prefix_paths = re.compile('^[0-9\s]+$')
+        regex_prefix_path_line1 = re.compile('.* from .*\([0-9a-fA-F.:]+\)')
+        regex_prefix_path_line2 = re.compile('\s+Origin')
+        regex_prefix_path_line3 = re.compile('\s+Last update:')
```

Comment:
it is better to give each line a meaning instead of 1,2,3...

Reply:
for example, line1 is source, line 2 is the attributes, line 3 is the update_time.

---
In reply to: [142343743](https://github.com/Azure/sonic-mgmt/pull/284#discussion_r142343743) [](ancestors = 142343743)